### ISSUE TITLE
MongoDB: `AND` operator giving different results based on order of queries

### ISSUE THREAD

**Issue by randbytes** [prisma/prisma#22970]
### Bug description
These queries giving different results, while native mongodb `$and` operator have predictable behaviour(results are equal, no matter the order)

### How to reproduce

1. Create similar model
```js
type Prop1 {}
type Prop2 {}
model MainModel {
  prop1  Prop1?
  prop2  Prop2?
}
```
2. Run these queries
```js
await prisma.model.count({where: {AND: [{prop1: {isNot: null}}, {prop2: {is: {name: "TEST"}}}]}})
await prisma.model.count({where: {AND: [{prop2: {is: {name: "TEST"}}}, {prop1: {isNot: null}}]}})
```
3. prop2 query ignored completely in first case

### Prisma information
```
generator client {
  provider = "prisma-client-js"
  binaryTargets = ["native", "linux-musl-openssl-3.0.x"]
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

type Prop1 {}
type Prop2 {}
model MainModel {
  prop1  Prop1?
  prop2  Prop2?
}
```

### Environment & setup

- OS: macos Ventura 13.6; Debian 12
- Database: MongoDB
- Node.js version:  v20.7.0

### Prisma Version

```
prisma                  : 5.9.1
@prisma/client          : 5.9.1
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v20.7.0
Query Engine (Node-API) : libquery-engine 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.9.0-32.23fdc5965b1e05fc54e5f26ed3de66776b93de64
Default Engines Hash    : 23fdc5965b1e05fc54e5f26ed3de66776b93de64
Studio                  : 0.497.0
```

**Comment by Weakky** [comment]
Hey @randbytes,

Would you mind trying with a newer version of Prisma and confirm whether the bug is still present?

Thank you 🙏

**Comment by SevInf** [comment]
I was not able to reproduce the issue in this specific case — it would've been really helpful to see the data in DB.

However, I 've confirmed almost identical https://github.com/prisma/prisma/issues/22263 that had a data sample. I also can confirmed that it was fixed in Prisma 5.11 (https://github.com/prisma/prisma-engines/pull/4754).

I am going to close this issue now, but feel free to comment on if it still happens to you on Prisma 5.11 or later.
